src/compiler/scala/tools/nsc/settings/MutableSettings.scala

```diff
@@ -213,9 +213,15 @@ class MutableSettings(val errorFn: String => Unit)
     }
 
   // a wrapper for all Setting creators to keep our list up to date
-  private def add[T <: Setting](s: T): T = {
-    allSettings += s
-    s
+  private[nsc] def add[T <: Setting](setting: T): T = {
+
+    allSettings.find(_.name == setting.name) match {
+      case Some(s) => allSettings -= s
+      case None =>
+    }
+
+    allSettings += setting
+    setting
```

Comment:
Setting doesn't have a useful hashCode or equals. Maybe ensure unique names this way:
```
diff --git a/src/compiler/scala/tools/nsc/settings/MutableSettings.scala b/src/compiler/scala/tools/nsc/settings/MutableSettings.scala
index a176552..b03b2a1 100644
--- a/src/compiler/scala/tools/nsc/settings/MutableSettings.scala
+++ b/src/compiler/scala/tools/nsc/settings/MutableSettings.scala
@@ -214,6 +214,7 @@ class MutableSettings(val errorFn: String => Unit)
 
   // a wrapper for all Setting creators to keep our list up to date
   private def add[T <: Setting](s: T): T = {
+    allSettings -= s
     allSettings += s
     s
   }
@@ -384,6 +385,13 @@ class MutableSettings(val errorFn: String => Unit)
     private var _deprecationMessage: Option[String] = None
     override def deprecationMessage = _deprecationMessage
     def withDeprecationMessage(msg: String): this.type = { _deprecationMessage = Some(msg) ; this }
+
+    override def hashCode = name.hashCode
+
+    override def equals(other: Any) = other match {
+      case s: Setting => name == s.name
+      case _ => false
+    }
   }

```
There are a lot of settings, so linear search suffers.